src/commoncode/fileutils.py

```diff
@@ -164,6 +170,64 @@ def get_temp_dir(base_dir, prefix=''):
     create_dir(base)
     return tempfile.mkdtemp(prefix=prefix, dir=base)
 
+
+def tree_checksum(tree_base_dir=None, _ignored=None):
+    """
+    Return a checksum computed from a file tree using the file paths,
+    size and last modified time stamps.
+    The purpose is to detect is there has been any modification to
+    source code or data files and use this as a proxy to verify the
+    cache consistency.
+
+    NOTE: this is not 100% fool proof but good enough in practice.
+    """
+    from commoncode import ignore
+
+    tree_base_dir = tree_base_dir or dirname(abspath(dirname(__file__)))
+
+    _ignored = _ignored or partial(
+        ignore.is_ignored,
+        ignores={'*.pyc': 'pyc files', '*~': 'temp gedit files', '*.swp': 'vi swap files'},
+        unignores={}
+    )
+
+    hashable = (pth + str(os.path.getmtime(pth)) + str(os.path.getsize(pth))
+                for pth in file_iter(tree_base_dir, ignored=_ignored))
+    return md5(''.join(sorted(hashable))).hexdigest()
+
+
+def create_cache_dir(dir_name, dev_mode=False):
+    """
+    Create directory `dir_name` and return it as an absolute path.
+
+    If `dev_mode` is True, the cache directory will be created in the
+    root directory of the source tree (e.g. the git repo root). If it
+    is False, it will be created in the current users home directory.
+
+    This location can be overridden by setting the environment variable
+    SCANCODE_TMP_DIR.
+    """
+    cache_path = build_cache_path(dir_name, dev_mode=dev_mode)
+    create_dir(cache_path)
+    return cache_path
+
+
+def build_cache_path(dir_name, dev_mode=False):
+    cache_root = os.environ.get(TMP_DIR_ENV)
+
+    if cache_root and not os.path.exists(cache_root):
+        raise OSError('{} is set to non-existent directory: {}'.format(TMP_DIR_ENV, cache_root))
```

Comment:
Why raise an error there? The directory should be created instead IMHO

Reply:
Setting the environment variable requires user intervention (unlike automatically picking cache dirs based on install method). Therefore I'd argue the user wants to be notified if there is a typo in the path, for example.
Creating a directory in the wrong place and proceeding with the scan might lead to filling up the wrong file system and could be time-consuming to diagnose.